services/notification_listener_test.go

```diff
@@ -276,3 +279,55 @@ func TestServices_RunLogTopic_ExpectedEventSignature(t *testing.T) {
 	expected := "0x06f4bf36b4e011a5c499cef1113c2d166800ce4013f6c2509cab1a0e92b83fb2"
 	assert.Equal(t, expected, services.RunLogTopic.Hex())
 }
+
+func TestHeadTracker_New(t *testing.T) {
+	t.Parallel()
+
+	store, cleanup := cltest.NewStore()
+	defer cleanup()
+	assert.Nil(t, store.Save(&models.BlockHeader{cltest.BigHexInt(1)}))
+	last := models.BlockHeader{cltest.BigHexInt(10)}
+	assert.Nil(t, store.Save(&last))
+	assert.Nil(t, store.Save(&models.BlockHeader{cltest.BigHexInt(2)}))
+
+	ht, err := services.NewHeadTracker(store)
+	assert.Nil(t, err)
+	assert.Equal(t, last.Number, ht.Get().Number)
+}
+
+func TestHeadTracker_Get(t *testing.T) {
+	t.Parallel()
+
+	store, cleanup := cltest.NewStore()
+	defer cleanup()
+	initial := models.BlockHeader{cltest.BigHexInt(1)}
+	assert.Nil(t, store.Save(&initial))
+
+	tests := []struct {
+		name      string
+		toSave    *models.BlockHeader
+		want      hexutil.Big
+		wantError bool
+	}{
+		// order matters
+		{"greater", &models.BlockHeader{cltest.BigHexInt(2)}, cltest.BigHexInt(2), false},
+		{"less than", &models.BlockHeader{cltest.BigHexInt(1)}, cltest.BigHexInt(2), false},
+		{"zero", &models.BlockHeader{cltest.BigHexInt(0)}, cltest.BigHexInt(2), true},
+		{"nil", nil, cltest.BigHexInt(2), true},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
```

Comment:
The order of the tests depends on the previous state of each past test. Why make them each separate test runs? It seems like you're relying on test pollution at that point, when they could just be considered one test.

Reply:
I thought it was practical to insert a sequence of block numbers, because it mirrored real world functionality by inserting a sequence, made the tests significant faster by not having to reinstantiate a Store, and didn't pollute because it was encapsulate in this one test deliberately, but shared among subtests.

The cost of manually configuring each scenario outweigh the reward.

Reply:
Discussed offline: I was suggesting simply removing `t.Run`.  Subtests were not a distinction I'd made, and they make debugging test failures significantly easier to track down. Sticking with it as is.